tests/testsuite/package.rs

```diff
@@ -1432,6 +1426,133 @@ edition = "2021"
+                [workspace.package]
+                edition = "2015"
+            "#,
+        )
+        .file(
+            "isengard/Cargo.toml",
+            r#"
+                [package]
+                name = "isengard"
+                edition.workspace = true
+                homepage = "saruman"
+                description = "saruman"
+                license = "MIT"
+            "#,
+        )
+        .file("isengard/src/lib.rs", "")
+    });
+    git::commit(&repo);
+
+    p.cargo("package --workspace --no-verify")
+        .with_stderr_data(str![[r#"
+[PACKAGING] isengard v0.0.0 ([ROOT]/foo/isengard)
+[PACKAGED] 5 files, [FILE_SIZE]B ([FILE_SIZE]B compressed)
+
+"#]])
+        .run();
+
+    // lockfile is untracked.
+    p.cargo("generate-lockfile").run();
+    p.cargo("package --workspace --no-verify")
+        .with_status(101)
+        .with_stderr_data(str![[r#"
+[ERROR] 1 files in the working directory contain changes that were not yet committed into git:
+
+Cargo.lock
+
+to proceed despite this and include the uncommitted changes, pass the `--allow-dirty` flag
+
+"#]])
+        .run();
+
+    // lockfile is tracked.
+    p.cargo("clean").run();
+    git::add(&repo);
+    git::commit(&repo);
+    p.cargo("package --workspace --no-verify")
+        .with_stderr_data(str![[r#"
+[PACKAGING] isengard v0.0.0 ([ROOT]/foo/isengard)
+[PACKAGED] 5 files, [FILE_SIZE]B ([FILE_SIZE]B compressed)
+
+"#]])
+        .run();
+
+    // Simulate that Cargo.lock had some outdated but SemVer compat dependency changes.
+    Package::new("dep", "1.0.0").publish();
+    Package::new("dep", "1.1.0").publish();
+    p.cargo("clean").run();
+    p.cargo("add dep@1").run();
+    git::add(&repo);
+    git::commit(&repo);
+    // make sure we have dep@1.1.0 in lockfile
+    assert_e2e().eq(
+        &p.read_lockfile(),
+        str![[r##"
+# This file is automatically @generated by Cargo.
+# It is not intended for manual editing.
+version = 4
+
+[[package]]
+name = "dep"
+version = "1.1.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "77d3d6a4f2203d590707cc803c94afbe36393bbdba757ef66986f39159eaab51"
+
+[[package]]
+name = "isengard"
+version = "0.0.0"
+dependencies = [
+ "dep",
+]
+
+"##]],
+    );
+    p.cargo("update dep --precise 1.0.0")
+        .with_stderr_data(str![[r#"
+[UPDATING] `dummy-registry` index
+[DOWNGRADING] dep v1.1.0 -> v1.0.0
+
+"#]])
+        .run();
+    p.cargo("package --workspace --no-verify")
+        .with_status(101)
+        .with_stderr_data(str![[r#"
+[ERROR] 1 files in the working directory contain changes that were not yet committed into git:
+
+Cargo.lock
+
+to proceed despite this and include the uncommitted changes, pass the `--allow-dirty` flag
+
+"#]])
+        .run();
+
+    // Now check if it is considered dirty when removed.
+    p.cargo("clean").run();
+    fs::remove_file(p.root().join("Cargo.lock")).unwrap();
+    p.cargo("package --workspace --no-verify")
```

Comment:
Might be some path normalization issues on Windows.

```
thread 'package::dirty_ws_lockfile_dirty' panicked at tests\testsuite\package.rs:1553:10:
test failed running `D:\a\cargo\cargo\target\debug\cargo.exe package --workspace --no-verify`
error: process exited with code 0 (expected 101)
--- stdout
--- stderr
   Packaging isengard v0.0.0 (D:\a\cargo\cargo\target\tmp\cit\t2239\foo\isengard)
    Updating `dummy-registry` index
    Packaged 5 files, 1.5KiB (986B compressed)
```

Reply:
Fixed by <https://github.com/rust-lang/cargo/compare/4c0e42dac2b05d52b52ef2435a9a25d73dd92c5e..f0907fca16a84101d69fc53272a955e8bb53d9fe>.

It was failing because removed files can't be canonicalized, so when comparing path prefix with a canonicalized `workdir` it will never pass.